### PR TITLE
bpo-30262: Deprecate sqlite Cache and Statement

### PR DESCRIPTION
**Pull request by palaviv**
https://bugs.python.org/issue30262

### CODE REVIEW

File: Lib/sqlite3/test/regression.py [hunk 1]
```diff
@@ -376,6 +376,14 @@ def CheckCommitCursorReset(self):
                 counter += 1
         self.assertEqual(counter, 3, "should have returned exactly three rows")
 
+    def CheckCacheDeprecationWarning(self):
```
Comment (berkerpeksag) [inline Lib/sqlite3/test/regression.py:402]: Can we move theses tests into a new `DeprecationTests` test case?

File: Modules/_sqlite/module.c [hunk 2]
```diff
@@ -320,6 +320,71 @@ static const IntConstantPair _int_constants[] = {
 };
 
 
+static PyObject *pysqlite_cache_deprecated_new(PyTypeObject *type, PyObject *args, PyObject *kwargs)
+{
+    pysqlite_Cache *self;
+
+    /* print deprecation warning */
```
Comment (berkerpeksag) [inline Modules/_sqlite/module.c]: Style nit: No need to add a comment for this.

File: Modules/_sqlite/module.c [hunk 3]
```diff
@@ -320,6 +320,71 @@ static const IntConstantPair _int_constants[] = {
...
+    if (PyErr_WarnEx(PyExc_DeprecationWarning,
+                     "Cache object has been deprecated",
```
Comment (berkerpeksag) [inline Modules/_sqlite/module.c]: Could you add "it won't be exposed in Python 3.9" or something similar as well?

File: Modules/_sqlite/module.c [hunk 4]
```diff
@@ -320,6 +320,71 @@ static const IntConstantPair _int_constants[] = {
...
+{
+    Py_TYPE(self)->tp_free((PyObject*)self);
+}
+
+
+static PyObject *pysqlite_statement_deprecated_new(PyTypeObject *type, PyObject *args, PyObject *kwargs)
+{
+    pysqlite_Statement *self;
+
+    /* print deprecation warning */
```
Comment (berkerpeksag) [inline Modules/_sqlite/module.c]: See my comments on `pysqlite_cache_deprecated_new()`.

File: Modules/_sqlite/module.c [hunk 5]
```diff
@@ -320,6 +320,71 @@ static const IntConstantPair _int_constants[] = {
...
+
+    self = (pysqlite_Statement *) pysqlite_StatementType.tp_alloc(&pysqlite_StatementType, 0);
+    if (!self) {
+        return NULL;
+    }
+
+    return (PyObject *)self;
+}
+
+/* classes for statement and cache deprecation */
```
Comment (berkerpeksag) [inline Modules/_sqlite/module.c]: Style nit: This can comment can be deleted.

File: Modules/_sqlite/module.c [hunk 6]
```diff
@@ -356,10 +423,10 @@ PyMODINIT_FUNC PyInit__sqlite3(void)
     PyModule_AddObject(module, "Connection", (PyObject*) &pysqlite_ConnectionType);
     Py_INCREF(&pysqlite_CursorType);
     PyModule_AddObject(module, "Cursor", (PyObject*) &pysqlite_CursorType);
-    Py_INCREF(&pysqlite_CacheType);
-    PyModule_AddObject(module, "Statement", (PyObject*)&pysqlite_StatementType);
-    Py_INCREF(&pysqlite_StatementType);
-    PyModule_AddObject(module, "Cache", (PyObject*) &pysqlite_CacheType);
+    Py_INCREF(&pysqlite_StatementDeprecatedType);
+    PyModule_AddObject(module, "Statement", (PyObject*)&pysqlite_StatementDeprecatedType);
```
Comment (berkerpeksag) [inline Modules/_sqlite/module.c]: Style nit: Since we already modified this line, we can adda space after `(PyObject*)` to make the styling consistent.

**Review by berkerpeksag — COMMENTED** [review]
We need to add a NEWS entry and a release note in `Doc/whatsnew/3.8.rst` (see the "Deprecated" section at https://docs.python.org/3.8/whatsnew/3.8.html#deprecated)

**Comment by berkerpeksag** [comment]
Oh, and please rebase `dont-export-sqlite-cache-and-statement2` branch :) Thanks!

**Comment by matrixise** [comment]
Hi @palaviv

Would you be interested to upgrade your PR to the last master?

Thank you

**Comment by palaviv** [comment]
Not needed after #1440.

**Pull request closed without merging**